src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorWidget.java

```diff
@@ -1098,48 +1118,102 @@ public void execute()
+            Marker marker = editor_.getSession().getMarker(markerId);
+
+            // The marker may have already been removed in response to
+            // a previous action.
+            if (marker == null)
+               continue;
+
+            String textRange = editor_.getSession().getTextRange(marker.getRange());
+            if (textRange != word)
+               newAnnotations.push(annotation.asAceAnnotation());
+            else
+               editor_.getSession().removeMarker(markerId);
+         }
+
+         editor_.getSession().setAnnotations(newAnnotations);
+         editor_.getRenderer().renderMarkers();
+      });
+   }
+
+   public void removeMarkers(Predicate<AceAnnotation> predicate)
+   {
+      // Defer this so other event handling can update anchors etc.
+      Scheduler.get().scheduleDeferred(() ->
+      {
+         JsArray<AceAnnotation> newAnnotations = JsArray.createArray().cast();
+
+         for (int i = 0; i < annotations_.size(); i++)
+         {
+            AnchoredAceAnnotation annotation = annotations_.get(i);
+            int markerId = annotation.getMarkerId();
+            Marker marker = editor_.getSession().getMarker(markerId);
+
+            // The marker may have already been removed in response to
+            // a previous action.
+            if (marker == null)
+               continue;
+
+            Debug.log(annotation.asAceAnnotation().text());
+            if (!predicate.test(annotation.asAceAnnotation()))
             {
-               AnchoredAceAnnotation annotation = annotations_.get(i);
-               int markerId = annotation.getMarkerId();
-               Marker marker = editor_.getSession().getMarker(markerId);
-               
-               // The marker may have already been removed in response to
-               // a previous action.
-               if (marker == null)
-                  continue;
-               
-               Range range = marker.getRange();
-               if (!range.contains(cursor))
-                  newAnnotations.push(annotation.asAceAnnotation());
-               else
-                  editor_.getSession().removeMarker(markerId);
+               newAnnotations.push(annotation.asAceAnnotation());
             }
-            
-            editor_.getSession().setAnnotations(newAnnotations);
-            editor_.getRenderer().renderMarkers();
-            
+            else
+               editor_.getSession().removeMarker(markerId);
          }
+
+         editor_.getSession().setAnnotations(newAnnotations);
+         editor_.getRenderer().renderMarkers();
+      });
+   }
+
+   public void removeMarkersAtCursorPosition()
+   {
+      // Defer this so other event handling can update anchors etc.
+      Scheduler.get().scheduleDeferred(() ->
+      {
+         Position cursor = editor_.getCursorPosition();
+         JsArray<AceAnnotation> newAnnotations = JsArray.createArray().cast();
+
+         for (int i = 0; i < annotations_.size(); i++)
```

Comment:
It looks like this code is more or less duplicated three times -- is it straightforward to share implementations?

Reply:
It was! I was able change up the predicate code in a way that was reusable. I wanted to refactor it last night but was a little stumped until I realized I could just use one of the functions already there.